### ISSUE TITLE
Proposal: Istio Envoy mTLS Private Key Protection with HSM

### ISSUE THREAD

**Issue by Chen-Xintong** [istio/istio#36296]
# TL; DR
Istio can better manage network traffic by sidecar auto-injection, and the enabling of mTLS can also protect the connection between services from being attacked and tampered, but there’s still vulnerabilities within the private key management in Istio, as the private key is exposed as clear text data inside the memory. The objective of this document is to provide a selectable choice to enable the PKCS# 11 standard inside Istio together with Hardware Security Module (HSM) to serve better security for key management within Istio. And enabling HSM to protect the mTLS private keys between workloads also provides users with a more secure choice. 

# Terminology
<html>
<body>
<b style="font-weight:normal;" id="docs-internal-guid-03ca96da-7fff-d473-8e04-c40987293b31"><div dir="ltr" style="margin-left:0pt;" align="left">

Abbreviation | Description
-- | --
HSM | Hardware security module
PKCS | Public-Key Cryptography Standard
PKCS# 11 | Public-Key Cryptography Standard. The PKCS# 11 standard defines a platform-independent API to cryptographic tokens, such as hardware security modules (HSM) and smart cards.

</div></b>
</body>
</html>

# Background 
## Current Workflow
Currently the use of mTLS protocol can provide a more secure and reliable connection, the flow diagram shows the current communication mode between sidecars and where the certificate will be used in this mesh.
<img width="1219" alt="flow1" src="https://user-images.githubusercontent.com/83945930/143997004-bd7eb380-6834-4c94-8eb6-3afe1c342162.png">

In the current workflow, the private keys which were  used for mTLS connection between sidecars were generated by Istio-Agent and issued to the sidecar, and the TLS private key is exposed as clear text data inside the memory of the pod. Although Istio's authentication policy can avoid most malicious third-party attacks, once a pod is compromised, the communication packages between the sidecar and other sidecars might be tampered with, which may lead to security vulnerabilities in the cluster.
 
In order to provide a way where the cluster can benefit from a pluggable third-party HSM backend and the envoy proxy can be extended to perform as a private key provider, we came up with this proposal. This design works for most use cases and provides an option that customers can choose to enable this option according to their needs. The workflow is shown in the diagram below.
<img width="1213" alt="flow2" src="https://user-images.githubusercontent.com/83945930/143997056-7510fcd6-efb9-4ce9-ba7f-fa46988c9c28.png">

Our changes are mainly in the process of generating CSR and key pair in Istio-Agent. The key point is to enable a selectable choice that the CSR and key pair can be generated outside Istio-Agent and protected by HSM backend. We introduced a global variable to enable the usage of external private key providers and in order to give customers a more flexible choice to adapt to all situations, if the external provider does not exist, the certificate and key pair will be generated by default flow.

A new annotation will be enabled when users choose to use HSM, the format of this annotation shown in the below, users can manually add this annotation to the configuration when deploying istio, the specific details will be explained later:

Annotations: sidecar.istio.io/external.provider: “true”

# Requirement

Istio-Agent certificate works as the mTLS connection certificate in Istio to provide security assurance for connections. However, the private keys exposed in memory provide more chances for attackers to exploit the system. By enabling PKCS# 11 standard as well as the help of HSM (which has several implementations), the mTLS private key can be saved and used more securely during the lifecycle of service mesh.

To enable the protection of mTLS by SoftHSM, workloads in the cluster need to support the corresponding HSM, and enable the corresponding option in bios settings, and then use HSM by function call via sidecar to the HSM.

# HSM & PKCS# 11 & SoftHSM
The hardware security module (HSM) is a special “trusted” environment which holds the happenings of a variety of cryptographic operations: key management, key exchange, encryption etc. It is also common to utilize HSM to enhance the security level for the software. The HSM can be implemented through either software or based on hardware, which totally depends on the users’ choice.

The PKCS # 11 standard defines a platform-independent API to cryptographic tokens, which is commonly used inside the HSM and actually called while doing cryptographic operations.

SoftHSM is an implementation of a cryptographic store accessible through a PKCS # 11 interface. It is used as a sample HSM implementation here.

# Design
This proposal tends to increase the security level of cryptographic operations and key management inside Istio. It involves modifying Istio-Agent to provide an advanced-level security to save the secrets (Istio-Agent mTLS private key) securely inside HSM (SoftHSM as sample) based on the PKCS# 11 standard, as an option (ENV is used to control, more things will get elaborated later). The workflow of generating CSR and certificate is shown in the diagram below.

![Flow4](https://user-images.githubusercontent.com/83945930/158095530-e89dee85-cd9b-4aca-afec-9db3707ad00f.png)

The main differences between our proposal and the current flow are:

1. Envoy's CSR is generated by Istio-agent with some HSM configs as extensions, and will be verified before signing.
2. The mTLS private key will no longer appear in plain-text, but will be fetched via handle and used by SoftHSM.
3. Istiod will do remote attestation before signing, then sign the CSR and add the attestation result to the certificate (optional).

## Istio-Agent Certificate generation/maintenance
To establish the mTLS communication between sidecars and gateway, Istiod or external CA like cert-manager takes the responsibility to issue the certificate for envoy, and this certificate is also a response of CSR from Istio-Agent. When Envoy requests its certificate and signing private key, it is Istio-Agent generates the corresponding CSR according to the current pod configuration, and signed mTLS certificate is stored in the memory by clear text data. Our proposal is to provide one more option in Istio-Agent to follow the PKCS# 11 standard and offload the key operations to HSM(SoftHSM) and provide higher security on the private keys. Every process will be discussed in the next few sections.

## Key generation
Two new approaches are provided for private key generation corresponding to the current. 

- Envoy will by default try to load the pre-conducted private key and save it into SoftHSM. 
- While there's no pre-conducted key ready, Envoy will request Istio-Agent for its private key and certificate, and Istio-Agent can call the PKCS11 APIs implemented inside the HSM(SoftHSM as sample) to generate key and the signed certificate, and then return to Envoy its key handle and cert. The key generation details are listed below:

1. Istio start Envoy.
2. Envoy requests its mTLS key from Istio-agent.
3. Istio-agent call SoftHSM API to generate key pair and CSR, and request CA for signing, some SoftHSM API may like below: 
4. Envoy uses this handle to fetch its private key via SoftHSM, and generates CSR calls Istio to sign it.

```
message Config {
    string hsm_library = 1;
    string usr_pin = 2;
    string so_pin = 3;
    string token_label = 4;
    string rsa_key_size = 5;
    // csr_config is a string composed of various configuration items
    // which contains if enable attestation and validator etc.
    string csr_config = 6;
}
```

For both cases, the output returned is a key handle instead of the plain-text key. 

## Key rotation

For some special scenarios, such as private key expired and someone regenerate the private key and request for a new certificate through JWT for his key, we have also implemented the following strategies: 

Both JWT and HSM Identity are required for regenerate the key and cert, and JWT can be a input for generate the HSM Identity so that a strong binding relationship will be formed between them. That is to say, only workloads trusted by HSM can regenerate new keys

## Key retrieval
While following the PKCS# 11 standard, id and label are used as the identifiers to fetch the private key. Either of them is usable to fetch the identical key. And the key that was retrieved is a handle instead of the plain-text data block. 

## Changes on deployment
People who want to enable this feature needs three extra steps on the deployment:

1. New Dockerfile containing the HSM libraries installed. The libraries may differ according to the HSM chosen by the deployer. We consider providing a sample based on SoftHSM first, since it is hardware agnostic. And more samples will be provided for other HSM implementations.
2. An ENV option called ‘USE_HSM_PROTECTION’ must be enabled to utilize the feature.
3. A customized istioOperator yaml file can be provided if there’s extra configurations needed for the deployment charts. Annotate the deployment as shown below.

```
selector:
    matchLabels:
      app: productpage
      version: v1
  template:
    metadata:
      annotations:
        sidecar.istio.io/external.provider: "true"
}
```

## Changes on usage
The changes will not be noticed by the end users, and the only difference is the deployment of istio with no change in the usage. End users can choose whether to use HSM to enhance the security of private keys when using the istioOperator install.

# Upgrade Support
Current in-place and Canary upgrade is supported for mesh which enables this feature.

# Test Plan
All original test cases will pass without any change.

New tests will be added to make sure when the env ‘USE_HSM_PROTECTION’ is set to ‘true’

- CSR can be successfully generated by envoy and is no different from the original.
- mTLS private key and cert generation flow thru SoftHSM will not break.
- mTLS private key recovery and rotation flow thru SoftHSM will not break.

And we need to have the HSM implementation integrated in the test framework.

**Affected product area (please put an X in all that apply)**

[ ] Docs
[ ] Installation
[x] Networking
[ ] Performance and Scalability
[ ] Extensions and Telemetry
[x] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

**Affected features (please put an X in all that apply)**

[ ] Multi Cluster
[ ] Virtual Machine
[ ] Multi Control Plane

**Additional context**

**Comment by howardjohn** [comment]
I don't immediately see the benefit. The claim here is that someone in the pod can access the memory of the envoy process and extract the private key. Sure... but if they are in the pod they can just generate a new certificate using the SA token. The HSM provides no security on that as far as I can tell. 

Am I missing something?

**Comment by Chen-Xintong** [comment]
> 

Hi Howard, thanks for your comment. As far as I know, the hardware security module (HSM) can play a special “trusted” environment role which holds the happenings of a variety of cryptographic operations: key management, key exchange, encryption etc. It is also common to utilize HSM to enhance the security level for the software. The entire process of generating and using the private key will not be exposed in the memory, so that the risk of being attacked can be completely avoided.

**Comment by Chen-Xintong** [comment]
> I don't immediately see the benefit. The claim here is that someone in the pod can access the memory of the envoy process and extract the private key. Sure... but if they are in the pod they can just generate a new certificate using the SA token. The HSM provides no security on that as far as I can tell.
> 
> Am I missing something?

Hi Howard, thanks for your comment. As far as I know, the hardware security module (HSM) can play a special “trusted” environment role which holds the happenings of a variety of cryptographic operations: key management, key exchange, encryption etc. It is also common to utilize HSM to enhance the security level for the software. The entire process of generating and using the private key will not be exposed in the memory, so that the risk of being attacked can be completely avoided.

**Comment by myidpt** [comment]
Hi Xintong, the point is that in the cert provision process, the K8s SA JWT is used as identity assertion file. This JWT is accessible through the pod persistent storage. Even though you use HSM to protect the workload X.509 key, the attacking vector associated with gaining access to the JWT still exists. The benefits of HSM is not obvious because the weak point is on the JWT.

**Comment by incfly** [comment]
I agree that if your concerned scenario is about envoy memory dump, JWT bearer token can be obtained and then be used to get workload key cert as well.

even though, @myidpt @howardjohn on the other hand, I think making key living in HSM might still have value. For example, this would disallow the attacker to first TAP and record the TLS traffic, and later trying to decrypt with memory dump-ed private key, since private key now never lives out of the HSM.

Their might be some other pure-compilance (not security related) value by just using HSM, but that's a different story...

Or @Chen-Xintong do you have some other setup/implicit details to overcome the concern about JWT token?  like using HSM as CA backend directly. or something else?

**Comment by Chen-Xintong** [comment]
> I agree that if your concerned scenario is about envoy memory dump, JWT bearer token can be obtained and then be used to get workload key cert as well.
> 
> even though, @myidpt @howardjohn on the other hand, I think making key living in HSM might still have value. For example, this would disallow the attacker to first TAP and record the TLS traffic, and later trying to decrypt with memory dump-ed private key, since private key now never lives out of the HSM.
> 
> Their might be some other pure-compilance (not security related) value by just using HSM, but that's a different story...
> 
> Or @Chen-Xintong do you have some other setup/implicit details to overcome the concern about JWT token? like using HSM as CA backend directly. or something else?

Hi Jianfei,

Thanks for your reply. As you said, on the one hand, HSM can be used as a trusted backend to store private keys securely. Even if TCP traffic is hijacked by a malicious third party, the traffic will not be decrypted because it is never possible to obtain the private key content in the HSM.

On the other hand, I am considering the possibility of using HSM to enhance the security of JWT. The initial idea is whether to enhance the certification process of applying for a new certificate with JWT. Thanks.

**Comment by Chen-Xintong** [comment]
Hi @howardjohn @myidpt and @incfly , I've considered your comments and I think the key point is that the JWT can be easily fetched once some third part malicious attacker can exec into the pod as root, so they can generate a new key pair and request Istiod signs a new certificate for the new key via JWT.

Since the JWT is usually mounted in the pod's memory which generally used as a pod's identity to request a certificate or used by services in the pod, we may not put the JWT directly into HSM, because if we just simply put it in the HSM, services in the pod must to get the JWT by a token, which may have more impact.

So I think we can fix this by adding an extra HSM attestation step before Istiod forwards the JWT to the api-server. When customers enable HSM feature, they can only get a new signed certificate when the HSM identity attestation successful and the JWT validity is also verified. Moreover, the general HSM will also have a corresponding verification mechanism. By verifying the token/pin or other characteristics of the HSM, it can be ensured that it is not a fake malicious third party applying for a certificate for his key pair.

**Comment by myidpt** [comment]
Hi @Chen-Xintong , Using HSM as an authority to attest the pod's identity sounds like the right direction. My question is how does the HSM authenticate the pod? I think it should be through a reliable out-of-band mechanism (not through means that the pod itself can conduct), for example, through the Kubelet assertion. In that case, the attacker won't be able to successfully provision certs from other random places (that can access HSM) using the victim pod's credentials.

**Comment by Chen-Xintong** [comment]
> Hi @Chen-Xintong , Using HSM as an authority to attest the pod's identity sounds like the right direction. My question is how does the HSM authenticate the pod? I think it should be through a reliable out-of-band mechanism (not through means that the pod itself can conduct), for example, through the Kubelet assertion. In that case, the attacker won't be able to successfully provision certs from other random places (that can access HSM) using the victim pod's credentials.

Hi @myidpt, for the HSM attestation, typically, the HSM will generate a identity in the trusted hardware area. And the HSM will generate completely different identities according to different input parameters, that is to say, HSM identities can form a mandatory binding relationship with key pair or SA Token. And the HSM have a corresponding absolutely trusted data center to verify the integrity of this identity and whether this identity matches the key pair. The process of HSM identity attestation and the JWT authentication can be as follows:

![mTLS_Attestation](https://user-images.githubusercontent.com/83945930/150128292-83bcacd8-ae2e-42e9-87b1-3cb7299714b6.png)

Based on the above principles, the attacker must obtain both the private key and the HSM identity to communicate with other workloads, which is impossible with the current HSM mechanism.

**Comment by Chen-Xintong** [comment]
Hi @myidpt @howardjohn, what do you think about this? Can this be an option for some admins?

**Comment by howardjohn** [comment]
How does the sidecar get the "HSM identity"? I am also a bit confused by the roll of Istiod here. is this using Istiod as an RA? Does Istiod need to know about HSMs?

**Comment by Chen-Xintong** [comment]
> How does the sidecar get the "HSM identity"? I am also a bit confused by the roll of Istiod here. is this using Istiod as an RA? Does Istiod need to know about HSMs?

This "HSM identity" is generated by HSM, which can be added to the CSR as an extension, and the Sidecar can get the encrypted “HSM identity" from the CSR extension. For the Istiod part, it don't need to know about HSM, just forward it to the trusted data center.

**Comment by Chen-Xintong** [comment]
For better understanding of our proposal, we draw 2 new graphs aiming the current design and our design difference ( the blue ones are new entities we will add).

> Current Flow
<img width="719" alt="currentflow" src="https://user-images.githubusercontent.com/83945930/156303469-4e77af1b-127a-463a-8204-f14f3d7d9c65.png">

> Our Proposal
<img width="724" alt="newflow" src="https://user-images.githubusercontent.com/83945930/156303608-f1cc2aef-c641-4f7b-b13f-3624db5e0595.png">


For the private key and Certificate generate process, the following diagram shows the respective roles of istio-agent, istiod, envoy and HSM:

Istio-agent distribute the related config about HSM and forward the CSR and HSM Identity form Envoy, Istiod can used as CA to request HSM attestation and certificate　signing, or as a RA to forward this CSR to a Kubernetes CA.

As you can see, even if the JWT is leaked by some attacker, he still can’t request his Certificate successfully because he will never get past the HSM Identity Attestation step.

We add HSM and Attestation Service based on current flow, thereby ensuring the security of the entire private key life cycle.

@howardjohn @myidpt @incfly What do you think about this?

**Comment by Chen-Xintong** [comment]
Hey @myidpt @howardjohn @incfly @justinpettit , according to the discussion in the Envoy community, I made some corresponding modifications to this proposal. The main improvement is that Istio-Agent can call HSM directly, which simplifies many steps and looks more reasonable. PTAL. Thanks!

**Comment by howardjohn** [comment]
For integration with SPIRE which is currently being implemented, the approach taken was to mount a custom UDS socket that is served by SPIREs server, and use that in place if istio-agent's SDS server.

I would recommend a similar approach here. Write a custom SDS service that uses this new HSM logic, and then use a custom CA that supports HSMs already. That way no changes are needed to istio-agent or Istiod.

WDYT?

**Comment by incfly** [comment]
> For integration with SPIRE which is currently being implemented, the approach taken was to mount a custom UDS socket that is served by SPIREs server, and use that in place if istio-agent's SDS server.

That sounds fair. We could make this new HSM-based  SDS server implementation available in a istio-ecosystem project initially. the website can list 

Other question, will the flow above can be adapted to major HSM vendor implementation? e.g. [AWS HSM](https://aws.amazon.com/cloudhsm/), [Azure HSM](https://azure.microsoft.com/en-us/services/azure-dedicated-hsm/), Thales HSM?

**Comment by Chen-Xintong** [comment]
I think this can make sense to implement a new "HSM SDS Server" responsible for the HSM key and CSR generation. And this server can be used by Istio-Agent through custom UDS. This part of the code can also be turned into a pluggable project and integrated with Istio, according to @incfly's suggestion, we can put it in a istio-ecosystem project and listed. 

And since the implementation of HSM servers of different vendors is relatively similar, if there are more HSMs of vendors who want to reuse this SDS Server in the future, we can also trying to design it in the form of framework + plugin.

According to my understanding, when HSM SDS Service is enabled (this mean custom UDS existed), the application and generation process for certificate and private key may be as follows:

![flow-sds drawio](https://user-images.githubusercontent.com/83945930/158329928-ba620154-628a-4f39-814e-9d8538cdef28.png)

@howardjohn @incfly  Does this make sense?

**Comment by fams** [comment]
One of the advantages of HSM is that even an attacker can generate connections or signatures inside the application; they cannot capture de private key to use it outside the controlled environment and without any trails. HSM can be used as a solution to call external services with mTLS requirements. It would be an excellent feature to add to the egress gateway.

**Comment by Chen-Xintong** [comment]
> One of the advantages of HSM is that even an attacker can generate connections or signatures inside the application; they cannot capture de private key to use it outside the controlled environment and without any trails. 

Exactly, the advantage of HSM is to ensure absolute security during the key life cycle.

> HSM can be used as a solution to call external services with mTLS requirements. It would be an excellent feature to add to the egress gateway.

For the usage scenario of gateway + HSM, the design and discussion are currently underway, which is also in our future scope, and this feature will also be supported.

**Comment by kmai** [comment]
How many calls/sec to the HSM do you expect to see on a normal cluster with around 10 nodes and 150 pods?

The first problem I could see is a bottleneck on your HSM delaying startup of pods, and it's expensive :p

**Comment by Chen-Xintong** [comment]
> How many calls/sec to the HSM do you expect to see on a normal cluster with around 10 nodes and 150 pods?
> 
> The first problem I could see is a bottleneck on your HSM delaying startup of pods, and it's expensive :p

We can support both local HSM and remote HSM.  Users can choose local HSM if they have performance requirements, and  users only need to do UDS communication once for key generation for local HSM.

**Comment by Chen-Xintong** [comment]
Hi @howardjohn,  according to our discussion [here](https://docs.google.com/document/d/1mtMgbNUr4y8_4uKVSNjJYrJS-lUttkNwEknDNBBLD8E/edit#), can we finalize this design so that we can start the functional implementation of HSM SDS Server.

**Comment by howardjohn** [comment]
My understanding is the HSM SDS server will be some external thing (can live in istio-ecosystem org if it needs a home), and there are no changes to Istio core? If that is the case then you can do whatever you want, even if I said no :-)

**Comment by Chen-Xintong** [comment]
> My understanding is the HSM SDS server will be some external thing (can live in istio-ecosystem org if it needs a home), and there are no changes to Istio core? If that is the case then you can do whatever you want, even if I said no :-)

Yes, we will follow the Spire integration flow and implement the functionality of HSM SDS Server, and it is better to put the code in istio-ecosystem org so that someone wants to use this function after code development, and also need to add a user manual in Istio maybe like [this ](https://github.com/istio/istio/pull/37948), thanks :)
